libs/langchain/langchain/schema/runnable/history.py

```diff
@@ -0,0 +1,254 @@
+from langchain.schema.runnable.config import RunnableConfig
+from langchain.schema.runnable.passthrough import RunnablePassthrough
+from langchain.schema.runnable.utils import (
+    ConfigurableFieldSpec,
+    get_unique_config_specs,
+)
+
+if TYPE_CHECKING:
+    from langchain.callbacks.tracers.schemas import Run
+    from langchain.schema.messages import BaseMessage
+
+MessagesOrDictWithMessages = Union[Sequence["BaseMessage"], Dict[str, Any]]
+
+
+class RunnableWithMessageHistory(RunnableBindingBase):
+    """
+    A runnable that manages chat message history for another runnable.
+
+    Base runnable must have inputs and outputs that can be converted to a list of
+        BaseMessages.
+
+    RunnableWithMessageHistory must always be called with a config that contains
+        {"configurable": {"thread_id": "<THREAD_ID>"}} and can optionally include
+        {"configurable": {"thread_id": "<THREAD_ID>", "user_id": "<USER_ID>"}}.
+
+    Example:
+        .. code-block:: python
+
+            from langchain.schema.prompts import ChatPromptTemplate, MessagesPlaceholder
+            from langchain.chat_models import ChatAnthropic
+            from langchain.schema.runnables import RunnableWithMessageHistory
+
+    """
+
+    factory: Callable[[str], BaseChatMessageHistory]
+    input_messages_key: Optional[str] = None
+    output_messages_key: Optional[str] = None
+    message_history_key: Optional[str] = None
+
+    def __init__(
+        self,
+        runnable: Runnable[
+            MessagesOrDictWithMessages,
+            Union[str, BaseMessage, MessagesOrDictWithMessages],
+        ],
+        factory: Callable[[str], BaseChatMessageHistory],
+        *,
+        input_messages_key: Optional[str] = None,
+        output_messages_key: Optional[str] = None,
+        message_history_key: Optional[str] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize RunnableWithMessageHistory.
+
+        Args:
+            runnable: The base Runnable to be wrapped.
+
+                Must take as input one of:
+                - A sequence of BaseMessages
+                - A dict with one key for all messages
+                - A dict with one key for the current input string/message(s) and
+                    a separate key for historical messages. If the input key points
+                    to a string, it will be treated as a HumanMessage in history.
+
+                Must return as output one of:
+                - A string which can be treated as an AIMessage
+                - A BaseMessage or sequence of BaseMessages
+                - A dict with a key for a BaseMessage or sequence of BaseMessages
+
+            factory: Function that returns a new BaseChatMessageHistory given a
+                session id. Should take a single parameter `session_id` which is a
+                string.
+            input_messages_key: Must be specified if the base runnable accepts a dict
```

Comment:
It looks like the code is falling back on `input` being a default, so this key doesn't have to be specified